org.eclipse.scanning.sequencer/src/org/eclipse/scanning/sequencer/nexus/SolsticeScanMonitor.java

```diff
@@ -0,0 +1,236 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * The scan points writer creates and writes to the unique keys and points
+ * datasets in a nexus file. The unique keys dataset can be used to track how far the scan has
+ * progressed.
+ * @author Matthew Dickie
+ */
+public class SolsticeScanMonitor extends AbstractScannable<Object> implements INexusDevice<NXcollection> {
+	
+	private static final Logger logger = LoggerFactory.getLogger(SolsticeScanMonitor.class);
+
+	// Nexus
+	private List<NexusObjectProvider<?>> nexusObjectProviders = null;
+	private NexusObjectWrapper<NXcollection> nexusProvider = null;
+	
+	// Writing Datasets
+	private ILazyWriteableDataset uniqueKeys = null;
+	private ILazyWriteableDataset scanFinished = null;
+
+	// State
+	private boolean malcolmScan = false;
+	private final ScanModel model;
+	
+	public SolsticeScanMonitor(ScanModel model) {
+		this.model = model;
+	}
+
+	public void setNexusObjectProviders(List<NexusObjectProvider<?>> nexusObjectProviders) {
+		this.nexusObjectProviders = nexusObjectProviders;
+	}
+	
+	public void setMalcolmScan(boolean malcolmScan) {
+		this.malcolmScan = malcolmScan;
+	}
+	
+	/* (non-Javadoc)
+	 * @see org.eclipse.dawnsci.nexus.INexusDevice#getNexusProvider(org.eclipse.dawnsci.nexus.NexusScanInfo)
+	 */
+	@Override
+	public NexusObjectProvider<NXcollection> getNexusProvider(NexusScanInfo info) {
+		
+		// Note: NexusScanFileManager relies on this method returning the same object each time
+		if (nexusProvider == null) {
+			NXcollection scanPointsCollection = createNexusObject(info);
+			nexusProvider = new NexusObjectWrapper<NXcollection>(GROUP_NAME_SOLSTICE_SCAN, scanPointsCollection);
+			nexusProvider.setPrimaryDataFieldName(FIELD_NAME_UNIQUE_KEYS);
+			nexusProvider.setAxisDataFieldNames(FIELD_NAME_UNIQUE_KEYS);
+		}
+		
+		return nexusProvider;
+	}
+
+	public NXcollection createNexusObject(NexusScanInfo info) {
+		final NXcollection scanPointsCollection = NexusNodeFactory.createNXcollection();
+		
+		// add a field for the scan rank
+		scanPointsCollection.setField(FIELD_NAME_SCAN_RANK, info.getRank());
+		
+		try {
+		    String cmd = ServiceHolder.getParserService().getCommand(model.getBean().getScanRequest(), true);
+			scanPointsCollection.setField(FIELD_NAME_SCAN_CMD,  cmd);
+		} catch (Exception ne) {
+			logger.debug("Unable to write scan command", ne);
+		}
+		
+		try {
+			List<?> models = model.getBean().getScanRequest().getCompoundModel().getModels();
+			String json = ServiceHolder.getMarshallerService().marshal(models);
+			scanPointsCollection.setField(FIELD_NAME_SCAN_MODELS, json);
+		} catch (Exception ne) {
+			logger.debug("Unable to write point models", ne);
+		}
+
+		// create the scan finished dataset and set the initial value to false
+//		scanFinished = scanPointsCollection.initializeFixedSizeLazyDataset(
+//				FIELD_NAME_SCAN_FINISHED, new int[] { 1 }, Dataset.INT32);
+		// TODO: workaround for bug in HD5 loader, do not set size limit 
```

Comment:
![INFO](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png 'Severity: INFO') Complete the task associated to this TODO comment. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonarqube.com/coding_rules#rule_key=squid%3AS1135)